code/datums/actions/mobs/charge.dm

```diff
@@ -0,0 +1,228 @@
+	/// The damage the charger does when bumping into something
+	var/charge_damage = 30
+	/// Bitflags for indications of charging
+	var/charge_indicator = TRUE
+	/// If we destroy objects while charging
+	var/destroy_objects = TRUE
+	/// Associative boolean list of chargers that are currently charging
+	var/list/charging = list()
+	/// Associative direction list of chargers that lets our move signal know how we are supposed to move
+	var/list/next_move_allowed = list()
+
+/datum/action/cooldown/mob_cooldown/charge/New(Target, delay, past, distance, speed, damage, destroy)
+	. = ..()
+	if(delay)
+		charge_delay = delay
+	if(past)
+		charge_past = past
+	if(distance)
+		charge_distance = distance
+	if(speed)
+		charge_speed = speed
+	if(damage)
+		charge_damage = damage
+	if(destroy)
+		destroy_objects = destroy
+
+/datum/action/cooldown/mob_cooldown/charge/Activate(atom/target_atom)
+	// start pre-cooldown so that the ability can't come up while the charge is happening
+	StartCooldown(10 SECONDS)
+	charge_sequence(owner, target_atom, charge_delay, charge_past)
+	StartCooldown()
+
+/datum/action/cooldown/mob_cooldown/charge/proc/charge_sequence(atom/movable/charger, atom/target_atom, delay, past)
+	do_charge(owner, target_atom, charge_delay, charge_past)
+
+/datum/action/cooldown/mob_cooldown/charge/proc/do_charge(atom/movable/charger, atom/target_atom, delay, past)
+	if(!target_atom || target_atom == owner)
+		return
+	var/chargeturf = get_turf(target_atom)
+	if(!chargeturf)
+		return
+	charger.setDir(get_dir(charger, target_atom))
+	var/turf/T = get_ranged_target_turf(chargeturf, charger.dir, past)
+	if(!T)
+		return
+	SEND_SIGNAL(owner, COMSIG_STARTED_CHARGE)
+	if(charge_indicator)
+		new /obj/effect/temp_visual/dragon_swoop/bubblegum(T)
+	RegisterSignal(charger, COMSIG_MOVABLE_BUMP, .proc/on_bump)
+	RegisterSignal(charger, COMSIG_MOVABLE_PRE_MOVE, .proc/on_move)
+	RegisterSignal(charger, COMSIG_MOVABLE_MOVED, .proc/on_moved)
+	charging[charger] = TRUE
+	DestroySurroundings(charger)
+	charger.setDir(get_dir(charger, target_atom))
+	if(charge_indicator)
+		var/obj/effect/temp_visual/decoy/D = new /obj/effect/temp_visual/decoy(charger.loc, charger)
+		animate(D, alpha = 0, color = "#FF0000", transform = matrix()*2, time = 3)
+	SLEEP_CHECK_DEATH(delay, charger)
+	var/distance = min(get_dist(charger, T), charge_distance)
+	for(var/i in 1 to distance)
+		// Prevents movement from the user during the charge
+		SLEEP_CHECK_DEATH(charge_speed, charger)
+		next_move_allowed[charger] = get_dir(charger, T)
+		step_towards(charger, T)
+		next_move_allowed.Remove(charger)
+	UnregisterSignal(charger, COMSIG_MOVABLE_BUMP)
+	UnregisterSignal(charger, COMSIG_MOVABLE_PRE_MOVE)
+	UnregisterSignal(charger, COMSIG_MOVABLE_MOVED)
+	charging.Remove(charger)
+	SEND_SIGNAL(owner, COMSIG_FINISHED_CHARGE)
+	return TRUE
+
+/datum/action/cooldown/mob_cooldown/charge/proc/on_move(atom/source, atom/new_loc)
+	var/expected_dir = next_move_allowed[source]
```

Comment:
```suggestion
/datum/action/cooldown/mob_cooldown/charge/proc/on_move(atom/source, atom/new_loc)
	SIGNAL_HANDLER

	var/expected_dir = next_move_allowed[source]
```
Also NONE of your signal procs have signal handlers as far as I can see, you should check them all